### PR TITLE
[feat] add `AES` Support to `DefaultCrypto` at `crypto-native` feature

### LINKED ISSUE #39: Use RustCrypto crates for AES-CBC/CTR
Following the [tweet](https://twitter.com/ShekoHex/status/1133399135233744897), I think you can use RustCrypto crates (in addition to `hmac` and `sha2`), namely: [`aes`](https://docs.rs/aes/), [`aes-ctr`](https://docs.rs/aes-ctr/) and [`block-modes`](https://docs.rs/block-modes/)

### PR DESCRIPTION
**Pull request by shekohex**
Hay @Michael-F-Bryan, I started work on that after @newpavlov helped us to find the needed rust crates for that.

By the way, i added a lot of comments in the code it self, so please review that too.

Closes #39 
Signed-off-by: Shady Khalifa <shekohex@gmail.com>

### CODE REVIEW

File: libsignal-protocol/src/crypto/mod.rs [hunk 1]
```diff
@@ -398,3 +399,46 @@ unsafe extern "C" fn internal_cipher(
         Err(e) => e.code(),
     }
 }
+
+#[cfg(all(test, feature = "crypto-native", feature = "crypto-openssl"))]
+mod crypto_tests {
+    use super::*;
+
+    #[test]
+    fn test_crypter() {
```
Comment (shekohex) [inline libsignal-protocol/src/crypto/mod.rs]: to run this test
use 
```bash
RUST_BACKTRACE=1 cargo test --no-fail-fast --features="crypto-native,crypto-openssl" test_crypter
```

File: libsignal-protocol/src/crypto/native.rs [hunk 2]
```diff
@@ -1,17 +1,112 @@
...
+    fn crypter(
+        &self,
+        mode: Mode,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError> {
+        let result = match (cipher, key.len()) {
+            (SignalCipherType::AesCtrNoPadding, 16) => {
```
Comment (shekohex) [inline libsignal-protocol/src/crypto/native.rs]: @Michael-F-Bryan i need to make sure of that, 
first i used a randomly generated nonce at every call using `self.fill_random(...)` but i found that using the `IV` here is a valid `nonce`, are you agree with that ?

File: libsignal-protocol/src/crypto/openssl.rs [hunk 3]
```diff
@@ -38,20 +39,36 @@ impl OpenSSLCrypto {
             (SignalCipherType::AesCbcPkcs5, 32) => Cipher::aes_256_cbc(),
             _ => unreachable!(),
         };
-
         let block_size = signal_cipher_type.block_size();
-        let mut result = Vec::with_capacity(data.len() + block_size);
```
Comment (shekohex) [inline libsignal-protocol/src/crypto/openssl.rs]: IMPORTANT :warning:

File: libsignal-protocol/src/crypto/openssl.rs [hunk 4]
```diff
@@ -38,20 +39,36 @@ impl OpenSSLCrypto {
...
+        // FIXME (@shekohex): Find why openssl has that behavior
+        let mut result = match cipher {
+            SignalCipherType::AesCtrNoPadding => vec![0u8; data.len()],
+            SignalCipherType::AesCbcPkcs5 => vec![0u8; data.len() + block_size],
+        };
         let mut crypter = Crypter::new(signal_cipher_type, mode, key, Some(iv))
             .map_err(|_e| InternalError::Unknown)?;
-
         crypter
             .update(data, &mut result)
```
Comment (shekohex) [inline libsignal-protocol/src/crypto/openssl.rs]: Reading throw source code of `crypter.update(...)` i found this
https://github.com/sfackler/rust-openssl/blob/ef86438a109f7f7683bde078563750a3ed02994d/openssl/src/symm.rs#L487-L505

especially this comment
```rust
    /// # Panics
    ///
    /// Panics for stream ciphers if `output.len() < input.len()`.
    ///
    /// Panics for block ciphers if `output.len() < input.len() + block_size`,
    /// where `block_size` is the block size of the cipher (see `Cipher::block_size`).
```

> stream ciphers (aka CTR), block ciphers (aka CBC), FYI, CTR has a block_size of 1, so no problem at all.

File: libsignal-protocol/src/crypto/mod.rs (hunk 1 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/crypto/mod.rs]: Can you move the `#[cfg(all(feature = "crypto-native", feature = "crypto-openssl"))]` to above the test? That way it's saying *this* test requires both native and OpenSSL crypto.

File: libsignal-protocol/src/crypto/mod.rs [hunk 5]
```diff
@@ -398,3 +399,46 @@ unsafe extern "C" fn internal_cipher(
...
+        native_crypto.fill_random(&mut key).unwrap();
+        native_crypto.fill_random(&mut iv).unwrap();
+
+        let cipher_text_native = native_crypto
+            .encrypt(SignalCipherType::AesCbcPkcs5, &key, &iv, &data)
+            .unwrap();
+        let cipher_text_openssl = openssl_crypto
+            .encrypt(SignalCipherType::AesCbcPkcs5, &key, &iv, &data)
+            .unwrap();
+        assert_eq!(cipher_text_native, cipher_text_openssl);
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/src/crypto/mod.rs]: Good idea making sure the native and OpenSSL implementations are identical!

File: libsignal-protocol/src/crypto/openssl.rs (hunk 3 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/crypto/openssl.rs]: Hmm, that trailing padding of 4 bytes is weird. Could it have something to do with being a block cipher and only emitting data in multiples of X bytes?

File: libsignal-protocol/src/crypto/openssl.rs (hunk 4 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/crypto/openssl.rs]: Do you think this is something we'll need to worry about? We could always document it and call it a day, otherwise there's probably a way to ensure the assumption is valid at a higher level...

File: libsignal-protocol/src/crypto/native.rs [hunk 6]
```diff
@@ -1,17 +1,112 @@
...
+                    .map_err(|_| InternalError::Unknown)?;
+                let buf = match mode {
+                    Mode::Encrypt => c.encrypt_vec(data),
+                    Mode::Decrypt => c
+                        .decrypt_vec(data)
+                        .map_err(|_| InternalError::Unknown)?,
+                };
+                buf
+            },
+            _ => unreachable!(),
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/src/crypto/native.rs]: We probably want to add more description to this `unreachable!()`. For example, `(cipher, size) => unreachable!("A combination of {} and {} doesn't make sense", cipher, size)`.

File: libsignal-protocol/src/crypto/native.rs (hunk 2 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/crypto/native.rs]: To be honest, I'm not familiar enough with the nuances of AES to know whether it's okay to use the provided IV as a nonce. Can we look at the `libsignal-protocol-c` source code and check whether our `crypter()` function will be provided with bytes that came from `crypto.fill_random()`?

File: libsignal-protocol/src/crypto/openssl.rs (hunk 3 above)
Comment (shekohex) [reply libsignal-protocol/src/crypto/openssl.rs]: turns out that, the crypter's `update` and `finalize` methods, return the `count` of bytes written in the buffer so far, so at the end of the `encryption/decryption` you should `truncate` the result buffer to that `count` of buffer.
by doing so, now it has the same result as `native` impl.

File: libsignal-protocol/src/crypto/mod.rs (hunk 5 above)
Comment (shekohex) [reply libsignal-protocol/src/crypto/mod.rs]: add both tests for `CBC/CTR` tests

File: libsignal-protocol/src/crypto/openssl.rs [hunk 7]
```diff
@@ -38,20 +39,24 @@ impl OpenSSLCrypto {
...
+        let mut count = crypter
             .update(data, &mut result)
             .map_err(|_e| InternalError::Unknown)?;
 
-        crypter
+        count += crypter
             .finalize(&mut result)
             .map_err(|_e| InternalError::Unknown)?;
-
+        result.truncate(count);
```
Comment (shekohex) [inline libsignal-protocol/src/crypto/openssl.rs]: here is the magic where we `truncate` the result buffer with overall `count` of bytes written to it, that drops all the trailing bytes at the end of the buffer

**Comment by shekohex** [comment]
Now, after adding a `DefaultCrypto` vs `OpenSSLCrypto` tests, we can see that it's almost painless to switch between the two with almost zero code change, and that make me so happy :tada:

**Comment by shekohex** [comment]
@Michael-F-Bryan feel free to merge this PR after reviewing it again.

Thank you :smile:
